### PR TITLE
refactor(cli/lsp): migrate from lspower back to tower-lsp

### PR DESCRIPTION
**Pull request by CGQAQ**
closed: #14056

### CODE REVIEW

**Comment by CGQAQ** [comment]
@kitsonk tower_lsp seems don't have [this method](https://docs.rs/lspower/1.5.0/src/lspower/lib.rs.html#1701), what should I do?

File: cli/lsp/language_server.rs [hunk 1]
```diff
@@ -2126,13 +2126,20 @@ impl Inner {
     }
   }
 
-  async fn request_else(
```
Comment (CGQAQ) [inline cli/lsp/language_server.rs:2129]: @kitsonk should I use this to replace request_else?

**Review by CGQAQ — COMMENTED** [review]
[is this ok?](https://github.com/denoland/deno/pull/14059/files#diff-20db93e8dfbff58c3024a4e247e639bf84fd5e8d2a447e8bbed9ee8942d688afL2129). @kitsonk

**Comment by CGQAQ** [comment]
I think it’s ready for review

File: cli/bench/lsp.rs [hunk 2]
```diff
@@ -6,12 +6,12 @@ use deno_core::serde_json;
 use deno_core::serde_json::json;
 use deno_core::serde_json::Value;
 use deno_core::url::Url;
-use lspower::lsp;
 use std::collections::HashMap;
 use std::path::Path;
 use std::time::Duration;
 use test_util::lsp::LspClient;
 use test_util::lsp::LspResponseError;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/bench/lsp.rs:14]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

This will avoid code churn.

File: cli/lsp/analysis.rs [hunk 3]
```diff
@@ -16,13 +16,13 @@ use deno_core::error::AnyError;
...
 use deno_core::serde_json::json;
 use deno_core::ModuleSpecifier;
-use lspower::lsp;
-use lspower::lsp::Position;
-use lspower::lsp::Range;
 use once_cell::sync::Lazy;
 use regex::Regex;
 use std::cmp::Ordering;
 use std::collections::HashMap;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/analysis.rs:23]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

Again, to reduce code churn.

File: cli/lsp/client.rs [hunk 4]
```diff
@@ -8,8 +8,8 @@ use deno_core::error::AnyError;
 use deno_core::futures::future;
 use deno_core::serde_json;
 use deno_core::serde_json::Value;
-use lspower::lsp;
-use lspower::lsp::ConfigurationItem;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/client.rs:11]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/code_lens.rs [hunk 5]
```diff
@@ -19,13 +19,13 @@ use deno_core::serde::Serialize;
...
 use deno_core::serde_json::json;
 use deno_core::ModuleSpecifier;
-use lspower::lsp;
 use once_cell::sync::Lazy;
 use regex::Regex;
 use std::cell::RefCell;
 use std::collections::HashSet;
 use std::rc::Rc;
 use std::sync::Arc;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/code_lens.rs:28]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/completions.rs [hunk 6]
```diff
@@ -20,10 +20,10 @@ use deno_core::serde::Serialize;
 use deno_core::url::Position;
 use deno_core::ModuleSpecifier;
 use import_map::ImportMap;
-use lspower::lsp;
 use once_cell::sync::Lazy;
 use regex::Regex;
 use std::sync::Arc;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/completions.rs:26]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/config.rs [hunk 7]
```diff
@@ -9,10 +9,10 @@ use deno_core::serde::Serialize;
 use deno_core::serde_json;
 use deno_core::serde_json::Value;
 use deno_core::ModuleSpecifier;
-use lspower::lsp;
 use std::collections::BTreeMap;
 use std::collections::HashMap;
 use std::sync::Arc;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/config.rs:15]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/diagnostics.rs [hunk 8]
```diff
@@ -711,10 +717,10 @@ impl DenoDiagnostic {
     }
   }
 
-  /// Given a reference to the code from an LSP diagnostic, determine if the
+  /// Given a reference to the code from an lsp_types diagnostic, determine if the
```
Comment (kitsonk) [inline cli/lsp/diagnostics.rs:720]: the comment is intentional, please revert the change

File: cli/lsp/documents.rs [hunk 9]
```diff
@@ -36,6 +35,7 @@ use std::path::Path;
 use std::path::PathBuf;
 use std::str::FromStr;
 use std::sync::Arc;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/documents.rs:38]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/language_server.rs [hunk 10]
```diff
@@ -87,7 +87,7 @@ pub(crate) struct Inner {
   /// A representation of metadata associated with specifiers in the DENO_DIR
   /// which is used by the language server
   cache_metadata: cache::CacheMetadata,
-  /// The LSP client that this LSP server is connected to.
+  /// The lsp_types client that this lsp_types server is connected to.
```
Comment (kitsonk) [inline cli/lsp/language_server.rs:90]: Please revert.

File: cli/lsp/language_server.rs [hunk 11]
```diff
@@ -115,13 +115,13 @@ pub(crate) struct Inner {
   pub(crate) maybe_import_map: Option<Arc<ImportMap>>,
   /// The URL for the import map which is used to determine relative imports.
   maybe_import_map_uri: Option<Url>,
-  /// A collection of measurements which instrument that performance of the LSP.
+  /// A collection of measurements which instrument that performance of the lsp_types.
```
Comment (kitsonk) [inline cli/lsp/language_server.rs:118]: Please revert

File: cli/lsp/language_server.rs [hunk 12]
```diff
@@ -115,13 +115,13 @@ pub(crate) struct Inner {
...
   performance: Arc<Performance>,
   /// A memoized version of fixable diagnostic codes retrieved from TypeScript.
   ts_fixable_diagnostics: Vec<String>,
   /// An abstraction that handles interactions with TypeScript.
   pub(crate) ts_server: Arc<TsServer>,
-  /// A map of specifiers and URLs used to translate over the LSP.
+  /// A map of specifiers and URLs used to translate over the lsp_types.
```
Comment (kitsonk) [inline cli/lsp/language_server.rs:124]: Please revert

File: cli/lsp/language_server.rs [hunk 13]
```diff
@@ -2126,13 +2126,20 @@ impl Inner {
     }
   }
 
-  async fn request_else(
+  // NOTE(CGQAQ): tower-lsp don't have [request_else](https://docs.rs/lspower/1.5.0/src/lspower/lib.rs.html#758)
```
Comment (kitsonk) [inline cli/lsp/language_server.rs:2129]: This note does not need to exist in the code... once migrated, no one will care.

File: cli/lsp/language_server.rs [hunk 14]
```diff
@@ -2726,12 +2740,12 @@ impl lspower::LanguageServer for LanguageServer {
     self.0.lock().await.rename(params).await
   }
 
-  async fn request_else(
+  // TODO(CGQAQ): tower_lsp don't have this method, should I just remove this one?
```
Comment (kitsonk) [inline cli/lsp/language_server.rs:2743]: No.  Just remove this comment.

File: cli/lsp/language_server.rs [hunk 15]
```diff
@@ -2770,7 +2784,7 @@ impl lspower::LanguageServer for LanguageServer {
   }
 }
 
-// These are implementations of custom commands supported by the LSP
+// These are implementations of custom commands supported by the lsp_types
```
Comment (kitsonk) [inline cli/lsp/language_server.rs:2787]: Please revert

File: cli/lsp/lsp_custom.rs [hunk 16]
```diff
@@ -2,7 +2,7 @@
 
 use deno_core::serde::Deserialize;
 use deno_core::serde::Serialize;
-use lspower::lsp;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/lsp_custom.rs:5]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/refactor.rs [hunk 17]
```diff
@@ -6,11 +6,11 @@
 use deno_core::serde::Deserialize;
 use deno_core::serde::Serialize;
 use deno_core::ModuleSpecifier;
-use lspower::lsp;
 use once_cell::sync::Lazy;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/refactor.rs:10]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/registries.rs [hunk 18]
```diff
@@ -31,12 +31,12 @@ use deno_graph::Dependency;
 use deno_runtime::deno_web::BlobStore;
 use deno_runtime::permissions::Permissions;
 use log::error;
-use lspower::lsp;
 use once_cell::sync::Lazy;
 use regex::Regex;
 use std::collections::HashMap;
 use std::path::Path;
 use std::path::PathBuf;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/registries.rs:39]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/text.rs [hunk 19]
```diff
@@ -6,14 +6,14 @@ use deno_core::serde_json::json;
...
-use lspower::jsonrpc;
-use lspower::lsp;
-use lspower::lsp::TextEdit;
 use std::collections::HashMap;
 use std::ops::Bound;
 use std::ops::RangeBounds;
 use text_size::TextRange;
 use text_size::TextSize;
+use tower_lsp::jsonrpc;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/text.rs:15]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/tsc.rs [hunk 20]
```diff
@@ -63,6 +60,9 @@ use text_size::TextSize;
 use tokio::sync::mpsc;
 use tokio::sync::oneshot;
 use tokio_util::sync::CancellationToken;
+use tower_lsp::jsonrpc::Error as LspError;
+use tower_lsp::jsonrpc::Result as LspResult;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/lsp/tsc.rs:65]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/lsp/tsc.rs [hunk 21]
```diff
@@ -163,7 +163,7 @@ struct AssetDocumentInner {
   maybe_navigation_tree: Option<Arc<NavigationTree>>,
 }
 
-/// An lsp representation of an asset in memory, that has either been retrieved
+/// An lsp_types representation of an asset in memory, that has either been retrieved
```
Comment (kitsonk) [inline cli/lsp/tsc.rs:166]: Please Revert

File: cli/lsp/tsc.rs [hunk 22]
```diff
@@ -108,7 +108,7 @@ impl TsServer {
         let mut started = false;
         while let Some((req, state_snapshot, tx, token)) = rx.recv().await {
           if !started {
-            // TODO(@kitsonk) need to reflect the debug state of the lsp here
+            // TODO(@kitsonk) need to reflect the debug state of the lsp_types here
```
Comment (kitsonk) [inline cli/lsp/tsc.rs:111]: Please revert

File: cli/lsp/tsc.rs [hunk 23]
```diff
@@ -2108,9 +2122,9 @@ impl CompletionEntry {
 
     let preselect = self.is_recommended;
     let use_code_snippet = settings.complete_function_calls
-      && (kind == Some(lsp::CompletionItemKind::FUNCTION)
-        || kind == Some(lsp::CompletionItemKind::METHOD));
-    // TODO(@kitsonk) missing from types: https://github.com/gluon-lang/lsp-types/issues/204
+      && (kind == Some(lsp_types::CompletionItemKind::FUNCTION)
+        || kind == Some(lsp_types::CompletionItemKind::METHOD));
+    // TODO(@kitsonk) missing from types: https://github.com/gluon-lang/lsp_types-types/issues/204
```
Comment (kitsonk) [inline cli/lsp/tsc.rs:2127]: Please revert

File: cli/lsp/tsc.rs [hunk 24]
```diff
@@ -2533,7 +2547,7 @@ fn op_exists(
   let state = state.borrow_mut::<State>();
   // we don't measure the performance of op_exists anymore because as of TS 4.5
   // it is noisy with all the checking for custom libs, that we can't see the
-  // forrest for the trees as well as it compounds any lsp performance
+  // forrest for the trees as well as it compounds any lsp_types performance
```
Comment (kitsonk) [inline cli/lsp/tsc.rs:2550]: please revert

File: cli/tests/integration/lsp_tests.rs [hunk 25]
```diff
@@ -8,15 +8,15 @@ use deno_core::serde_json;
...
 use deno_core::url::Url;
-use lspower::lsp;
 use pretty_assertions::assert_eq;
 use std::collections::HashSet;
 use std::fs;
 use tempfile::TempDir;
 use test_util::deno_exe_path;
 use test_util::http_server;
-use test_util::lsp::LspClient;
+use test_util::lsp::{LspClient, LspResponseError};
```
Comment (kitsonk) [inline cli/tests/integration/lsp_tests.rs:17]:
```suggestion
use test_util::lsp::LspClient;
use test_util::lsp::LspResponseError;
```

File: cli/tests/integration/lsp_tests.rs [hunk 26]
```diff
@@ -8,15 +8,15 @@ use deno_core::serde_json;
...
 use test_util::testdata_path;
+use tower_lsp::lsp_types;
```
Comment (kitsonk) [inline cli/tests/integration/lsp_tests.rs:19]:
```suggestion
use tower_lsp::lsp_types as lsp;
```

File: cli/tests/integration/lsp_tests.rs [hunk 27]
```diff
@@ -1765,7 +1782,7 @@ fn lsp_hover_dependency() {
   );
 }
 
-// This tests for a regression covered by denoland/deno#12753 where the lsp was
+// This tests for a regression covered by denoland/deno#12753 where the lsp_types was
```
Comment (kitsonk) [inline cli/tests/integration/lsp_tests.rs:1785]: Please revert

File: cli/lsp/language_server.rs (hunk 13 above)
Comment (CGQAQ) [reply cli/lsp/language_server.rs:2129]: ok :)

File: Cargo.lock [hunk 28]
```diff
@@ -4126,7 +4145,7 @@ checksum = "9fb5b7c98597bf41d1503ca4039be5445fd02e7aa381ae520d1c78a8a370f7f5"
 dependencies = [
  "ahash",
  "base64 0.13.0",
- "dashmap",
+ "dashmap 4.0.2",
```
Comment (bartlomieju) [inline Cargo.lock:4148]: This PR introduces a lot of duplicated dependencies with multiple versions. I suggest to hold off on landing it before we can deduplicate them

File: cli/lsp/language_server.rs [hunk 29]
```diff
@@ -2126,13 +2126,16 @@ impl Inner {
     }
   }
 
-  async fn request_else(
+  async fn execute_command(
```
Comment (kitsonk) [inline cli/lsp/language_server.rs:2129]: Actually, this solution isn't acceptable, because it is a significant breaking change with all clients.  We can't merge this.  Clients use custom methods, not "workspace/executeCommand" that is a totally different LSP API.  Basically every client would break if we merged this. :sad:

**Review by kitsonk — CHANGES_REQUESTED** [review]
We can't break the API this way.  We need tower_lsp to support _other_ requests, versus trying to repurpose an existing lsp method.

**Comment by kitsonk** [comment]
tower_lsp added the feature for custom server requests here: https://github.com/ebkalderon/tower-lsp/pull/313.  We need to implement it that way.

**Comment by CGQAQ** [comment]
Ok, I'll change the tower-lsp from 0.15.1 to 0.16
